tests/integrationv2/test_buffered_send.py

```diff
@@ -0,0 +1,209 @@
+]
+
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", ALL_TEST_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("other_provider", [S2N, OpenSSL, GnuTLS], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", [S2N], ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", PROTOCOLS, ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("buffer_size", [SEND_BUFFER_SIZE_HUGE, SEND_BUFFER_SIZE_MIN], ids=get_parameter_name)
+@pytest.mark.parametrize("fragment_pref", [None, "--prefer-low-latency", "--prefer-throughput"], ids=get_parameter_name) # Test various fragment prefrences
+def test_s2n_buffered_send_all_settings_server(managed_process, cipher, other_provider, provider, protocol, certificate, buffer_size, fragment_pref):
+    port = next(available_ports)
+
+    random_bytes = data_bytes(SEND_DATA_SIZE)
+    client_options = ProviderOptions(
+        mode=Provider.ClientMode,
+        port=port,
+        cipher=cipher,
+        data_to_send=random_bytes,
+        insecure=True,
+        protocol=protocol)
+
+    if other_provider is S2N:
+        client_options.extra_flags = ['--buffered-send', buffer_size]
+
+    server_options = ProviderOptions(
+        mode=Provider.ServerMode,
+        port=port,
+        cipher=cipher,
+        data_to_send=None,
+        insecure=True,
+        protocol=protocol)
+ 
+    server_options.extra_flags = ['--buffered-send', buffer_size]
+    if fragment_pref is not None:
+        server_options.extra_flags.append(fragment_pref)
+
+    server_options.key = certificate.key
+    server_options.cert = certificate.cert
+    server_options.cipher = None
+
+    server = managed_process(provider, server_options, timeout=5)
+    client = managed_process(other_provider, client_options, timeout=5)
+
+    for results in client.get_results():
+        results.assert_success()
+
+    for results in server.get_results():
+        results.assert_success()
+        if provider is S2N: random_bytes in results.stdout
+
+
+@pytest.mark.uncollect_if(func=invalid_test_parameters)
+@pytest.mark.parametrize("cipher", TLS13_CIPHERS, ids=get_parameter_name)
+@pytest.mark.parametrize("other_provider", [S2N, OpenSSL, GnuTLS], ids=get_parameter_name)
+@pytest.mark.parametrize("provider", [S2N], ids=get_parameter_name)
+@pytest.mark.parametrize("protocol", [Protocols.TLS13], ids=get_parameter_name)
+@pytest.mark.parametrize("certificate", ALL_TEST_CERTS, ids=get_parameter_name)
+@pytest.mark.parametrize("buffer_size", SEND_BUFFER_SIZES, ids=get_parameter_name)
+@pytest.mark.parametrize("fragment_pref", [None, "--prefer-low-latency", "--prefer-throughput"], ids=get_parameter_name) # Test various fragment prefrences
+def test_s2n_buffered_send_all_sizes_server(managed_process, cipher, other_provider, provider, protocol, certificate, buffer_size, fragment_pref):
+    port = next(available_ports)
```

Comment:
Yeah I agree.

As it turns out this isn't a problem anymore after limiting the ciphers to a subset.

None the less I think investigating this future is worth while -- the test infrastructure here comes in multiple layers and works in parallel. The `next(available_ports)` is handing out ports based on the thread you are in. Using port 0 might be better way to make this work -- we can change that by changing the implementation of available_port -- but it does seem to work for reasonably sized tests.

Reply:
Can you create an issue for the port problem, if an issue doesn't exist yet?

Also, just to be clear in case that's not just a typo, you should be limiting the certificates to a subset, NOT the ciphers. All ciphers should still be used.

Reply:
Yes it was a typo. Only limiting the certificates not the ciphers.